### PR TITLE
Add functionality for passing 'source' to pkg_ref

### PR DESCRIPTION
**Pull request by elimillera**
Issue #177. Add options to set 'source' when creating package references.

## Description
This PR adds functionality to pass a character vector to the `pkg_ref` to specify a source.

## Example
```
ref <- pkg_ref("dplyr", source = "pkg_remote")
```

## Tests added
Tests were added to make sure the right option is selected by default and can be changed. Also tests support for multiple packages with different sources.

### CODE REVIEW

**Comment by elong0527** [comment]
add relevant issue #175

**Comment by elong0527** [comment]
Current scope: define three sources:

1. `pkg_remote`: current CRAN release 
2. `pkg_installed`: package installed and identified by `.libPath`

Future development:

1. how to specify specific package version and its dependency version.   
2. add `pkg_source`: a folder that contain user saved packages.

**Comment by elimillera** [comment]
@elong0527 @dgkf 
Most of the development should be good here. 'source' and 'lib.loc' can be passed to `pkg_ref` to specify a source and package library. I am going to add some documentation and this should be good. If you could try to break this that would be appreciated!

File: R/pkg_ref_class.R [hunk 1]
```diff
@@ -5,24 +5,31 @@
...
 #' package. As data is needed for assessing a package's risk, this metadata
 #' populates fields within the package reference object.
 #'
-#' The \code{pkg_ref} S3 subclasses are used extensively for divergent metdata
+#' The \code{pkg_ref} S3 subclasses are used extensively for divergent metadata
 #' collection behaviors dependent on where the package was discovered. Because
 #' of this, there is a rich hierarchy of subclasses to articulate the different
 #' ways package information can be found.
 #'
+#' A source argument can be passed using the `source` argument. This will
```
Comment (elong0527) [inline R/pkg_ref_class.R:20]: Please add examples under `@examples` to illustrate the usage of `source` argument. 

Please descrbie possible value for `source` argument.

File: R/pkg_ref_class.R [hunk 2]
```diff
@@ -122,69 +141,110 @@ as_pkg_ref.pkg_ref <- function(x, ...) {
...
+           path <- find.package(x, lib.loc = lib.loc)
+           version <- utils::packageVersion(x, lib.loc = dirname(path))
+
+           new_pkg_ref(x,
+                       version = version,
+                       path = path,
+                       source = "pkg_install")
+         },
+         pkg_source = {
+
```
Comment (elong0527) [inline R/pkg_ref_class.R:167]: Could we provide a more informative error message when `source` is not available? 

Below is what currently have. 

```
> ref5 <- pkg_ref("testthat", source = "pkg_source")
Error in read.dcf(file.path(x, "DESCRIPTION")) : 
  cannot open the connection
In addition: Warning message:
In read.dcf(file.path(x, "DESCRIPTION")) :
 
 Error in read.dcf(file.path(x, "DESCRIPTION")) : 
  cannot open the connection 
```

Comment (elong0527) [reply R/pkg_ref_class.R:167]: Similarly, I feel we shall we report warning/error if a package is not in a source? 

```
> ref5 <- pkg_ref("testthat2", source = "pkg_cran_remote")
> ref5
<pkg_cran_remote, pkg_ref> testthat2 v
$repo
  character(0)
$source
  [1] "pkg_cran_remote"
$version
  character(0)
$name
  [1] "testthat2"
$archive_release_dates...
$bug_reports...
$bug_reports_host...
$bug_reports_url...
$downloads...
$license...
$news_urls...
$source_control_url...
$web_url...
$website_urls...
```

File: R/pkg_ref_class.R [hunk 3]
```diff
@@ -122,69 +141,110 @@ as_pkg_ref.pkg_ref <- function(x, ...) {
...
 
-      return(new_pkg_ref(x,
-        version = version,
-        path = path,
-        source = "pkg_install"))
+#' Determine the intended source of a new package
+#'
+#' @param x Package name or path to package
+#' @param source type of source passed in `pkg_ref`
+#' @return one of c('pkg_install', 'pkg_install', 'pkg_cran_remote', 'pkg_bioc_remote')
```
Comment (elong0527) [inline R/pkg_ref_class.R:204]: will `pkg_missing` also be one of return?

**Review by elong0527 — CHANGES_REQUESTED** [review]
Good work! We shall be able to merge it after we align the comments below.

File: R/pkg_ref_class.R (hunk 1 above)
Comment (elimillera) [reply R/pkg_ref_class.R:20]: Examples added

File: R/pkg_ref_class.R (hunk 2 above)
Comment (elimillera) [reply R/pkg_ref_class.R:180]: I've added nice warnings for these and change the source type to 'pkg_missing". Tests have been added to check for this behavior

File: R/pkg_ref_class.R (hunk 3 above)
Comment (elimillera) [reply R/pkg_ref_class.R]: Updated

**Review by dgkf — APPROVED** [review]
Just had a chance to do a cursory look and didn't spot anything that sparked any concerns from me. Thanks @elimillera

**Review by elong0527 — APPROVED** [review]

**Pull request merged**